lib/tasks/result_monitor.rake

```diff
@@ -32,6 +32,51 @@ class MonitorPipelineResults
         LogUtil.log_backtrace(exception)
       end
     end
+
+    # "stalled uploads" are not pipeline jobs, but they fit in here better than
+    # anywhere else.
+    begin
+      MonitorPipelineResults.alert_stalled_uploads!
+    rescue => exception
+      LogUtil.log_err_and_airbrake("Failed to alert on stalled uploads: #{exception.message}")
+      LogUtil.log_backtrace(exception)
+    end
+
+    begin
+      MonitorPipelineResults.fail_stalled_uploads!
+    rescue => exception
+      LogUtil.log_err_and_airbrake("Failed to fail stalled uploads: #{exception.message}")
+      LogUtil.log_backtrace(exception)
+    end
+  end
+
+  def self.fail_stalled_uploads!
+    # Delay determined by doubling the time after which an alert is sent to admins.
+    Sample
+      .stalled_uploads(6.hours)
+      .each { |sample| sample.fail_local_upload!("Stalled for more than 6 hours") }
+  end
+
+  def self.alert_stalled_uploads!
+    # Delay determined based on query of historical upload times, where 80%
+    # of successful uploads took less than 3 hours by client_updated_at.
+    samples = Sample.stalled_uploads(3.hours).where(upload_error: nil)
+    if samples.empty?
+      return
+    end
+
+    samples.each { |sample| sample.update(upload_error: Sample::UPLOAD_ERROR_LOCAL_UPLOAD_STALLED) }
```

Comment:
same comment about update_all

Reply:
With update_all, you'll also avoid triggering all these before/after_save callbacks that have no business being triggered here: https://github.com/chanzuckerberg/idseq-web/blob/master/app/models/sample.rb#L78

Reply:
Same comment re performance.

Thanks for pointing out those callbacks. It seems like if they are not generally applicable, they should be restricted to certain types of events.

Reply:
Updated